### DISCUSSION TITLE
Custom training with custom yaml file

### DISCUSSION THREAD

**Discussion by CyrusAddy** [ultralytics/ultralytics#21263] [Q&A]
I am using this code for training my custom yaml (yolov8n_p2_out.yaml) file from scratch, but I still have some layers freezing. What could be the cause?
yolo detect train data=/workspace/data/customdata.yaml pretrained=False model=/workspace/data/yolov8n_p2_out.yaml freeze=0 epochs=100 imgsz=640 project=Yolo_Training name=yolov8n_P2_out_coal_miner_1

**Comment by UltralyticsAssistant** [comment]
👋 Hello @CyrusAddy, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for comprehensive guides, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as many common questions are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

If this is a custom training ❓ Question, please include as much detail as possible—such as your dataset, relevant YAML files, and training logs. Make sure to review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for more support! For real-time chat, head over to [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, visit [Discourse](https://community.ultralytics.com/), or check out our [Subreddit](https://www.reddit.com/r/Ultralytics/) to connect with others.

## Upgrade

Upgrade to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue hasn’t already been resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your discussion and assist you soon!

**Reply by glenn-jocher** [reply]
When training from scratch with `pretrained=False`, you typically shouldn't need to specify `freeze=0` at all since no layers should be frozen by default. The issue might be in your custom YAML file or how the model is being initialized.

Can you share your training logs to see which specific layers are being frozen? Also, try removing the `freeze=0` parameter entirely since it shouldn't be needed when building from scratch, and consider testing with the Python API to have more control:

```python
from ultralytics import YOLO

model = YOLO("/workspace/data/yolov8n_p2_out.yaml")  # build from YAML
results = model.train(data="/workspace/data/customdata.yaml", epochs=100, imgsz=640)
```

**Reply by CyrusAddy** [reply]
wandb: Currently logged in as: cyrusaddy90 (spot34) to https://api.wandb.ai/. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.21.0
wandb: Run data is saved locally in /ultralytics/wandb/run-20250702_201227-5pnlt95k
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run yolov8n_P2_out_coal_miner_12
wandb: ⭐️ View project at https://wandb.ai/spot34/Yolo_Training
wandb: 🚀 View run at https://wandb.ai/spot34/Yolo_Training/runs/5pnlt95k
**Freezing layer 'model.14.dfl.conv.weight'**
AMP: running Automatic Mixed Precision (AMP) checks...
AMP: checks passed ✅
train: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1712.8±746.3 MB/s, size: 337.1 KB)
train: Scanning /workspace/data/coal_miner/labels/train.cache... 12461 images, 0 backgrounds, 0 corrupt: 100%|██████████| 12461/12461 [00:00<?, ?it/s]
albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01, method='weighted_average', num_output_channels=3), CLAHE(p=0.01, clip_limit=(1.0, 4.0), tile_grid_size=(8, 8))
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 812.1±466.3 MB/s, size: 292.8 KB)
val: Scanning /workspace/data/coal_miner/labels/val.cache... 3145 images, 0 backgrounds, 0 corrupt: 100%|██████████| 3145/3145 [00:00<?, ?it/s]
Plotting labels to Yolo_Training/yolov8n_P2_out_coal_miner_12/labels.jpg...
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 35 weight(decay=0.0), 40 weight(decay=0.0005), 39 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to Yolo_Training/yolov8n_P2_out_coal_miner_12

**Reply by CyrusAddy** [reply]
I am using Kubernetes so what is a suitable CLI command for this:

from ultralytics import YOLO

model = YOLO("/workspace/data/yolov8n_p2_out.yaml")  # build from YAML
results = model.train(data="/workspace/data/customdata.yaml", epochs=100, imgsz=640)

**Reply by glenn-jocher** [reply]
Looking at your logs, only one layer is being frozen: `model.14.dfl.conv.weight` - this is actually expected behavior. The DFL (Distribution Focal Loss) layer is typically frozen during training as it's a mathematical operation layer, not a learnable parameter layer.

For the CLI equivalent in Kubernetes, use:

```bash
yolo detect train data=/workspace/data/customdata.yaml model=/workspace/data/yolov8n_p2_out.yaml epochs=100 imgsz=640
```

Note that I removed `pretrained=False` since it's not needed when building from a YAML file, and removed `freeze=0` since the single frozen layer is expected behavior.
